aw_query/functions.py

```diff
@@ -47,31 +48,58 @@ def _verify_variable_is_type(variable, t):
 """
     Declarations
 """
-query2_functions = {}  # type: Dict[str, TQueryFunction]
+functions = {}  # type: Dict[str, TQueryFunction]
 
 
-def q2_function(f):
+def q2_function(transform_func=None):
     """
     Decorator used to register query functions.
 
     Automatically adds mock arguments for Datastore and TNamespace
     if not in function signature.
     """
+
+    def h(f):
+        sig = signature(f)
+        # If function lacks docstring, use docstring from underlying function in aw_transform
+        if transform_func and transform_func.__doc__ and not f.__doc__:
+            f.__doc__ = ".. note:: Documentation automatically copied from underlying function `aw_transform.{func_name}`\n\n{func_doc}".format(func_name=transform_func.__name__, func_doc=transform_func.__doc__)
+
+        @wraps(f)
+        def g(datastore: Datastore, namespace: TNamespace, *args, **kwargs):
+            # Remove datastore and namespace argument for functions that don't need it
+            args = (datastore, namespace, *args)
+            if TNamespace not in (sig.parameters[p].annotation for p in sig.parameters):
+                args = (args[0], *args[2:])
+            if Datastore not in (sig.parameters[p].annotation for p in sig.parameters):
+                args = (args[1:])
+            return f(*args, **kwargs)
+
+        fname = f.__name__
+        if fname[:3] == "q2_":
```

Comment:
We should raise an Exception if this is not the case, every function with that annotation should be prefixed with "q2_"

Reply:
Why? What's wrong with, in theory, registering functions from `aw_transform` directly?